Does this plugin support javacc 4.0v?

Hi,

 Now we are migrating our legacy product over there we need to use javacc 4.0 version to convert .jjt files

dependencies {
    javacc 'net.java.dev.javacc:javacc:4.0'
}
when we use the above dependencies it throws below error

> Task :llt:compileJjtree FAILED
Java Compiler Compiler Version 4.0 (Tree Builder)
(type "jjtree" with no arguments for help)
Exception in thread "main" java.lang.NullPointerException
	at org.javacc.parser.Options.setCmdLineOption(Unknown Source)
	at org.javacc.jjtree.JJTree.main(Unknown Source)
	at org.javacc.jjtree.Main.main(Unknown Source)

FAILURE: Build failed with an exception.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -24,3 +24,7 @@ jobs:
     - name: Run build script
       run: ./gradlew build
       shell: bash
+    - name: checkstyle
+      uses: jwgmeligmeyling/checkstyle-github-action@master
+      with:
+        path: '**/reports/checkstyle/*.xml'

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -25,6 +25,8 @@ jobs:
       run: ./gradlew build
       shell: bash
     - name: checkstyle
-      uses: jwgmeligmeyling/checkstyle-github-action@master
+      uses: dbelyaev/action-checkstyle@v0.7.4
       with:
-        path: '**/reports/checkstyle/*.xml'
+        github_token: ${{ secrets.github_token }}
+        reporter: github-check
+        level: warning

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -28,5 +28,6 @@ jobs:
       uses: dbelyaev/action-checkstyle@v0.7.4
       with:
         github_token: ${{ secrets.github_token }}
+        checkstyle_config: 'config/checkstyle/checks.xml'
         reporter: github-check
         level: warning

--- a/subprojects/plugin/build.gradle
+++ b/subprojects/plugin/build.gradle
@@ -54,7 +54,11 @@ jar {
 checkstyle {
     configFile = file("${rootProject.projectDir}/config/checkstyle/checks.xml")
 }
-
+tasks.withType(Checkstyle) {
+    reports {
+        xml.required = true
+    }
+}
 spotbugs {
     toolVersion = '4.2.1'
     effort = "min"

--- a/subprojects/plugin/src/acceptanceTest/java/javacc/compilation/CompilationSteps.java
+++ b/subprojects/plugin/src/acceptanceTest/java/javacc/compilation/CompilationSteps.java
@@ -46,7 +46,7 @@ public class CompilationSteps {
         ensureGivens();
 
         String[] defaultArguments = new String[] {
-            "--debug", "--stacktrace", "--project-dir", projectDirectory.getAbsolutePath(), "-b", "build.gradle"
+            "--info", "--stacktrace", "--project-dir", projectDirectory.getAbsolutePath(), "-b", "build.gradle"
         };
 
         String[] allArguments = ArrayUtils.addAll(defaultArguments, arguments);
@@ -67,7 +67,7 @@ public class CompilationSteps {
 
     public void thenAssertOutputDirectoryExists(String outputDirectory) {
         this.outputDirectory = new File(projectDirectory, outputDirectory);
-        assertTrue(this.outputDirectory.exists());
+        assertExists(this.outputDirectory);
     }
 
     public void thenAssertOutputDirectoryDoesNotExists(String outputDirectory) {
@@ -75,12 +75,16 @@ public class CompilationSteps {
     }
 
     public void andAssertFileWasGenerated(String filename) {
-        assertTrue((new File(outputDirectory, filename)).exists());
+        assertExists(new File(outputDirectory, filename));
+    }
+
+    private void assertExists(File file) {
+        assertTrue("File " + file + " should exist", file.exists());
     }
 
     public void andAssertFileExistsButWasNotGenerated(String filename) throws IOException {
         File javaFile = new File(outputDirectory, filename);
-        assertTrue(javaFile.exists());
+        assertExists(javaFile);
 
         String fileContent = FileUtils.readFileToString(javaFile);
         assertTrue(fileContent.contains("public static final boolean IS_CUSTOM = true;"));

--- a/subprojects/plugin/src/acceptanceTest/java/javacc/compilation/ThePluginIsCompatibleWithAllSpportedJavaccVersions.java
+++ b/subprojects/plugin/src/acceptanceTest/java/javacc/compilation/ThePluginIsCompatibleWithAllSpportedJavaccVersions.java
@@ -1,0 +1,67 @@
+package javacc.compilation;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import org.junit.Test;
+
+public class ThePluginIsCompatibleWithAllSpportedJavaccVersions {
+
+    @Test
+    public void givenVersion4WhenExecuteAllTasksFilesAreGenerated() throws URISyntaxException, IOException {
+        givenVersionWhenExecuteAllTasksFilesAreGenerated("net.java.dev.javacc:javacc:4.0", true);
+    }
+
+    @Test
+    public void givenVersion5WhenExecuteAllTasksFilesAreGenerated() throws URISyntaxException, IOException {
+        givenVersionWhenExecuteAllTasksFilesAreGenerated("net.java.dev.javacc:javacc:5.0", true);
+    }
+
+    @Test
+    public void givenVersion6WhenExecuteAllTasksFilesAreGenerated() throws URISyntaxException, IOException {
+        givenVersionWhenExecuteAllTasksFilesAreGenerated("net.java.dev.javacc:javacc:6.1.2", true);
+    }
+
+    @Test
+    public void givenVersion7WhenExecuteAllTasksFilesAreGenerated() throws URISyntaxException, IOException {
+        givenVersionWhenExecuteAllTasksFilesAreGenerated("net.java.dev.javacc:javacc:7.0.12", true);
+    }
+
+    @Test
+    public void givenVersion8WhenExecuteAllTasksFilesAreGenerated() throws URISyntaxException, IOException {
+        givenVersionWhenExecuteAllTasksFilesAreGenerated("org.javacc.generator:java:8.0.1", false);
+    }
+
+    public void givenVersionWhenExecuteAllTasksFilesAreGenerated(String javaccDependency, boolean simpleNode)
+        throws URISyntaxException, IOException {
+        CompilationSteps steps = new CompilationSteps();
+
+        steps.givenAProjectNamed("multiprojectBuildWithVersion");
+        steps.withArguments("clean", ":subprojects:subproject1:compileJavacc", ":subprojects:subproject1:jjdoc",
+            "-PjavaccDep=" + javaccDependency).execute();
+
+        String buildDirectory = "subprojects" + File.separator + "subproject1" + File.separator + "build" + File.separator + "generated";
+
+        steps.thenAssertOutputDirectoryExists(buildDirectory + File.separator + "javacc");
+        steps.andAssertFileWasGenerated("Hello.java");
+        steps.andAssertFileWasGenerated("HelloTokenManager.java");
+        steps.andAssertFileWasGenerated("SimpleCharStream.java");
+        steps.andAssertFileWasGenerated("TokenMgrError.java");
+        steps.andAssertFileWasGenerated("HelloConstants.java");
+        steps.andAssertFileWasGenerated("ParseException.java");
+        steps.andAssertFileWasGenerated("Token.java");
+
+        steps.thenAssertOutputDirectoryExists(buildDirectory + File.separator + "jjtree");
+        steps.andAssertFileWasGenerated("JJTreeOutputTest.jj");
+        steps.andAssertFileWasGenerated("HelloTreeConstants.java");
+        steps.andAssertFileWasGenerated("JJTHelloState.java");
+        steps.andAssertFileWasGenerated("Node.java");
+        if (simpleNode) {
+            steps.andAssertFileWasGenerated("SimpleNode.java");
+        }
+        steps.thenAssertOutputDirectoryExists(buildDirectory + File.separator + "jjdoc");
+        steps.andAssertFileWasGenerated("MyParser.html");
+    }
+
+}

--- a/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/build.gradle
+++ b/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/build.gradle
@@ -1,0 +1,7 @@
+allprojects{
+    group = 'org.javacc.plugin.gradle.acceptanceTests'
+
+    repositories {
+        mavenCentral()
+    }
+}

--- a/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/settings.gradle
+++ b/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/settings.gradle
@@ -1,0 +1,3 @@
+include 'subprojects:subproject1', 'subprojects:subproject2'
+
+rootProject.name = 'multiprojectBuildWithJJTree'

--- a/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/build.gradle
+++ b/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/build.gradle
@@ -1,0 +1,21 @@
+plugins {
+    id 'java'
+    id 'eclipse'
+    id 'org.javacc.plugin.javacc-gradle'
+}
+
+sourceCompatibility = 1.8
+version = '1.0'
+jar {
+    manifest {
+        attributes 'Implementation-Title': 'Gradle Quickstart', 'Implementation-Version': archiveVersion
+    }
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    javacc project.findProperty("javaccDep")
+}

--- a/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/src/main/javacc/MyParser.jj
+++ b/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/src/main/javacc/MyParser.jj
@@ -1,0 +1,27 @@
+options
+    { IGNORE_CASE = true; }
+    PARSER_BEGIN(MyParser)
+
+public class MyParser {
+    public static void main(String[] args) throws ParseException {
+        MyClass mc = new MyClass();
+    }
+}
+    PARSER_END(MyParser)
+
+SKIP:
+    {
+    " "
+    }
+
+    TOKEN:
+    {
+<ANYTHING:  ~[]>
+    }
+
+
+    void production():
+    {}
+    {
+    (<ANYTHING>)+
+    }

--- a/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/src/main/jjtree/JJTreeOutputTest.jjt
+++ b/subprojects/plugin/src/acceptanceTest/resources/javacc/compilation/projects/multiprojectBuildWithVersion/subprojects/subproject1/src/main/jjtree/JJTreeOutputTest.jjt
@@ -1,0 +1,17 @@
+PARSER_BEGIN(Hello)
+package AST;
+public class Hello {}
+PARSER_END(Hello)
+
+SKIP : { " " | "\t" | "\n" | "\r" }
+
+TOKEN :
+{
+    <ABS: "abs">
+}
+
+SimpleNode Start() : {}
+{
+  <ABS>
+  { return jjtThis; }
+}

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/compiler/JavaccSourceFileCompiler.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/compiler/JavaccSourceFileCompiler.java
@@ -75,7 +75,6 @@ public class JavaccSourceFileCompiler implements SourceFileCompiler {
     private ProgramArguments buildProgramArguments(File inputDirectory, RelativePath inputRelativePath) {
         ProgramArguments commandLineArguments = new ProgramArguments(argumentsProvidedByTask);
         commandLineArguments.addFilename(inputRelativePath.getFile(inputDirectory).getAbsolutePath());
-
         return programInvoker.augmentArguments(inputDirectory, inputRelativePath, commandLineArguments);
     }
 

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/AbstractProgramInvoker.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/AbstractProgramInvoker.java
@@ -3,10 +3,13 @@ package org.javacc.plugin.gradle.javacc.programexecution;
 import java.io.File;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.stream.Collectors;
 
 import org.gradle.api.Action;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
 import org.gradle.process.ExecResult;
 import org.gradle.process.JavaExecSpec;
 
@@ -37,5 +40,39 @@ abstract class AbstractProgramInvoker implements ProgramInvoker {
         Constructor<? extends Action<JavaExecSpec>> constructor = executorType.getConstructor(Configuration.class, ProgramArguments.class);
 
         return constructor.newInstance(classpath, arguments);
+    }
+
+    protected void addCodeGenerator(ProgramArguments augmentedArguments) {
+        Integer[] version = getJavaccVersion();
+        if (version.length >  0 && version[0] >= 8) {
+            augmentedArguments.add("CODE_GENERATOR", "java");
+        }
+    }
+
+    protected Integer[] getJavaccVersion() {
+        String version = "";
+        try {
+            Configuration configuration = project.getConfigurations().getByName("javacc");
+            for (Dependency dependency : configuration.getAllDependencies()) {
+                String id = dependency.getGroup() + ":" + dependency.getName();
+                if (dependency.getVersion() != null
+                    && ("net.java.dev.javacc:javacc".equals(id)
+                    || "org.javacc.generator:java".equals(id))) {
+                    version = dependency.getVersion();
+                }
+            }
+        } catch (RuntimeException ex) {
+            ex.printStackTrace();
+        }
+        return Arrays.stream(version.split("\\.")).map(this::safeParseInt).toArray(Integer[]::new);
+    }
+
+    private int safeParseInt(String num) {
+        try {
+            return Integer.parseInt(num);
+        } catch (RuntimeException ignore) {
+            // non-numeric version
+        }
+        return 0;
     }
 }

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/AbstractProgramInvoker.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/AbstractProgramInvoker.java
@@ -4,7 +4,6 @@ import java.io.File;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.util.Arrays;
-import java.util.stream.Collectors;
 
 import org.gradle.api.Action;
 import org.gradle.api.Project;
@@ -14,6 +13,9 @@ import org.gradle.process.ExecResult;
 import org.gradle.process.JavaExecSpec;
 
 abstract class AbstractProgramInvoker implements ProgramInvoker {
+
+    protected static final int VERSION_4 = 4;
+    protected static final int VERSION_8 = 8;
     protected final File tempOutputDirectory;
 
     private final Project project;
@@ -44,7 +46,7 @@ abstract class AbstractProgramInvoker implements ProgramInvoker {
 
     protected void addCodeGenerator(ProgramArguments augmentedArguments) {
         Integer[] version = getJavaccVersion();
-        if (version.length >  0 && version[0] >= 8) {
+        if (version.length >  0 && version[0] >= VERSION_8) {
             augmentedArguments.add("CODE_GENERATOR", "java");
         }
     }

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JavaccProgramInvoker.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JavaccProgramInvoker.java
@@ -20,6 +20,7 @@ public class JavaccProgramInvoker extends AbstractProgramInvoker {
     public ProgramArguments augmentArguments(File inputDirectory, RelativePath inputRelativePath, ProgramArguments arguments) {
         ProgramArguments augmentedArguments = new ProgramArguments(arguments);
         augmentedArguments.add("OUTPUT_DIRECTORY", inputRelativePath.getFile(tempOutputDirectory).getParentFile().getAbsolutePath());
+        addCodeGenerator(augmentedArguments);
         return augmentedArguments;
     }
 

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JjtreeProgramInvoker.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JjtreeProgramInvoker.java
@@ -4,7 +4,6 @@ import java.io.File;
 
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
-import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.file.RelativePath;
 
 /**
@@ -21,8 +20,8 @@ public class JjtreeProgramInvoker extends AbstractProgramInvoker {
 
     private String getOutputDirectoryArgName() {
         Integer[] version = getJavaccVersion();
-        if (version[0] == 4 && version[1] <=0) {
-           return "OUTPUT_DIRECTORY";
+        if (version[0] == VERSION_4 && version[1] <= 0) {
+            return "OUTPUT_DIRECTORY";
         }
         return "JJTREE_OUTPUT_DIRECTORY";
     }

--- a/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JjtreeProgramInvoker.java
+++ b/subprojects/plugin/src/main/java/org/javacc/plugin/gradle/javacc/programexecution/JjtreeProgramInvoker.java
@@ -4,6 +4,7 @@ import java.io.File;
 
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.file.RelativePath;
 
 /**
@@ -11,15 +12,26 @@ import org.gradle.api.file.RelativePath;
  */
 public class JjtreeProgramInvoker extends AbstractProgramInvoker {
     public static final String SUPPORTED_FILE_SUFFIX = ".jjt";
+    private final String outputDirectoryArgName;
 
     public JjtreeProgramInvoker(Project project, Configuration classpath, File tempOutputDirectory) {
         super(project, classpath, tempOutputDirectory, JjtreeExecutorAction.class);
+        outputDirectoryArgName = getOutputDirectoryArgName();
+    }
+
+    private String getOutputDirectoryArgName() {
+        Integer[] version = getJavaccVersion();
+        if (version[0] == 4 && version[1] <=0) {
+           return "OUTPUT_DIRECTORY";
+        }
+        return "JJTREE_OUTPUT_DIRECTORY";
     }
 
     @Override
     public ProgramArguments augmentArguments(File inputDirectory, RelativePath inputRelativePath, ProgramArguments arguments) {
         ProgramArguments augmentedArguments = new ProgramArguments(arguments);
-        augmentedArguments.add("JJTREE_OUTPUT_DIRECTORY", inputRelativePath.getFile(tempOutputDirectory).getParentFile().getAbsolutePath());
+        augmentedArguments.add(outputDirectoryArgName, inputRelativePath.getFile(tempOutputDirectory).getParentFile().getAbsolutePath());
+        addCodeGenerator(augmentedArguments);
         return augmentedArguments;
     }
 

--- a/subprojects/plugin/src/test/java/org/javacc/plugin/gradle/javacc/programexecution/JavaExecFactory.java
+++ b/subprojects/plugin/src/test/java/org/javacc/plugin/gradle/javacc/programexecution/JavaExecFactory.java
@@ -1,17 +1,17 @@
 package org.javacc.plugin.gradle.javacc.programexecution;
 
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
 import org.gradle.api.internal.provider.DefaultProperty;
 import org.gradle.api.internal.provider.PropertyHost;
 import org.gradle.api.provider.Property;
 import org.gradle.process.JavaExecSpec;
 
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
-
 public class JavaExecFactory {
     public static JavaExecSpec createSpec() {
         JavaExecSpec spec = mock(JavaExecSpec.class);
-        Property<String> mainClass = new DefaultProperty<>(mock(PropertyHost.class),String.class);
+        Property<String> mainClass = new DefaultProperty<>(mock(PropertyHost.class), String.class);
         when(spec.getMainClass()).thenReturn(mainClass);
         return spec;
     }

